### ISSUE TITLE
Public page of a user returns 404

### ISSUE THREAD

**Issue by verymilan** [tootsuite/mastodon#13402]
Hi there, [i've tried to get help in the forum already, but noone replied](https://discourse.joinmastodon.org/t/one-specific-public-user-profile-returns-404/2574) - i have one particular (local) user whose public profile page returns a 404 error. This is the second case i am aware of, the first one i wrote about in the forum left the instance due to this in the meantime.

No moderative actions have been done to this user, it is also no private profile.
No other problems (usersearch, interacting, ...) are known.
I didn't see errors in my log(s) (which is not set to debug yet).

The first user reported that sometimes their profile would show up, but most of the time not.

The issue was last seen in v3.1.3. The previously affected user said that it happened somewhere around the v3.1.0 update.

**Comment by ThibG** [comment]
I suspect for some reason something raises an `ActiveRecord::RecordNotFound` exception when accessing that user. If so, you may be able to get a stacktrace by commenting out https://github.com/tootsuite/mastodon/blob/v3.1.3/app/controllers/application_controller.rb#L28, restarting `mastodon-web` (`systemctl reload mastodon-web`), reloading the user page (which should now return a 500 rather than a 404), while looking at mastodon-web's logs (`journalctl -u mastodon-web --follow`).

**Comment by mshayden** [comment]
I have encountered a profile with the same issue on my instance after upgrading to 3.1.3 (I skipped over 3.1.0).  I have also reported this in the forum:

https://discourse.joinmastodon.org/t/users-public-profile-missing-404-the-page-you-are-looking-for-isnt-here/2672

I commented out the line in question and received this stack trace in my mastodon-web logs.  Something to do with locales?  I will investigate further and see if I can work around (maybe it is fixed by changing the locale or adding it to my server?).  In any case there is an exception that is not handled properly here.  I don't have experience with Ruby but if the fix is simple I'll try to submit a patch.

`
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473] method=GET path=/@<REDACTED> format=html controller=AccountsController action=show status=404 error='ActiveRecord::RecordNotFound: ActiveRecord::RecordNotFound' duration=183.04 view=0.00 db=49.68
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473]
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473] ActiveRecord::RecordNotFound (ActiveRecord::RecordNotFound):
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473]
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473] app/models/concerns/account_finder_concern.rb:8:in `find_local!'
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473] app/controllers/concerns/account_owned_concern.rb:20:in `set_account'
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473] app/controllers/concerns/localized.rb:18:in `block in set_locale'
Apr 17 12:56:44 app-0004 bundle[12550]: [ce1b5989-83e9-4181-8c21-bba4f4e36473] app/controllers/concerns/localized.rb:17:in `set_locale'
`

**Comment by ThibG** [comment]
@mshayden the error doesn't have anything to do with locales, it's just that the whole request handling is wrapped by a function that makes sure the correct locale is in use.

The error not being caught is thanks to you commenting the line, which helps us investigate.

From the rails console (`bundle exec rails c`), can you check the following commands (with the username you redacted instead of `FOO`)?

`Account.find_by(domain: nil, username: 'FOO')`
`Account.find_local!('FOO')`

I'm suspecting your PostgreSQL indexes might be corrupted, but I might be wrong

**Comment by mshayden** [comment]
Thanks for the explanation--first time I've dived into any sort of Ruby code much less Mastodon!

find_by() returns:

`=> #<Account username: "<REDACTED>", domain: nil, secret: "", private_key: "-----BEGIN RSA PRIVATE KEY-----\n<REDACTED>", public_key: "-----BEGIN PUBLIC KEY-----\n<REDACTED>", remote_url: "", salmon_url: "", hub_url: "", created_at: "2019-09-04 19:41:06", updated_at: "2020-04-09 20:25:54", note: "<REDACTED>", display_name: "<REDACTED>", uri: "", url: nil, avatar_file_name: "97d2ba51cc9a942a.jpg", avatar_content_type: "image/jpeg", avatar_file_size: 41592, avatar_updated_at: "2020-04-09 20:25:53", header_file_name: nil, header_content_type: nil, header_file_size: nil, header_updated_at: nil, avatar_remote_url: nil, subscription_expires_at: nil, locked: false, header_remote_url: "", last_webfingered_at: nil, inbox_url: "", outbox_url: "", shared_inbox_url: "", followers_url: "", protocol: "ostatus", id: 58406, memorial: false, moved_to_account_id: nil, featured_collection_url: nil, fields: [<REDACTED>], actor_type: "Person", discoverable: true, also_known_as: nil, silenced_at: nil, suspended_at: nil, trust_level: nil, hide_collections: nil>`

find_local!() returns:

`Traceback (most recent call last):
        2: from (irb):2
        1: from app/models/concerns/account_finder_concern.rb:8:in `find_local!'
ActiveRecord::RecordNotFound (ActiveRecord::RecordNotFound)`

**Comment by ThibG** [comment]
Yeah, this seems to be a corrupted index issue…
I'd try running the following in PostgreSQL: `REINDEX INDEX index_accounts_on_username_and_domain_lower`

**Comment by mshayden** [comment]
Definitely index corruption.  There are a lot of duplicate username/domain records in the accounts table preventing this from happening successfully (always just a second record, not more thankfully).  I am trying to clean these up, though with the corrupt index it is playing mind games with me.  The common pattern is that the created_at of all the offending records started right after I did the database migration to 3.1.3.  I am not sure what in the code would corrupt the index--I think it was most likely corrupted by something else previously and the migration just triggered this to happen maybe?

**Comment by ThibG** [comment]
The duplicate username/domain records are most probably because of the corrupted index. Unfortunately, I have no idea what could have caused the index corruption… it shouldn't have been something from Mastodon itself…

**Comment by mshayden** [comment]
The whole incident started when my docker-based install crashed and docker-compose refused to restart.  I killed the docker setup with fire, spun up a fresh instance installed from source and attached the (still-intact) volumes on my SAN to the new instance, basically following the documentation for migrating an instance to a new server.  That is why I think it is highly likely it was not the code that corrupted the database--it was something to do with docker crashing.

To fix the issue I did the following:

Step 1: stopped the mastodon-* services

Step 2: ran the following query (risky! do a DB dump first to be safe!):

`delete from accounts where id in (select new.id from (select id, username, domain from accounts where created_at < '2020-04-13') as old inner join (select id, username, domain from accounts where created_at >= '2020-04-13') as new on old.username = new.username and old.domain = new.domain);`

Step 3: I tried the REINDEX and deleting the duplicates that were still there one at a time until it worked lol! (there were just a small handful of remote accounts where both copies were created after  2020-04-13).  This step probably isn't required if you write a smarter query, but I assumed there wouldn't be any duplicates of very new accounts

Step 4: restarted mastodon-* services.

It is interesting that the index in question wasn't even defined as unique when I first started my instance.  I think the UNIQUE constraint was added later on, but I'm pretty sure that was not the culprit because UNIQUE was added earlier than 3.x

Hopefully this helps others fix the issue if they run into it.
